### PR TITLE
feat: add option for throwing exception on non-successful API request

### LINKED ISSUE #775: [Suggestion] SendGridClient.SendEmailAsync should throw when a non-success response is received.
Currently, if I call `SendGridClient.SendEmailAsync` and the API returns an error, it doesn't throw an exception. I have to check the response (which is made harder than necessary by #774) to see what happened. It's counter-intuitive; if the method doesn't throw, anyone would expect that the mail was successfully sent (or at least that the request was accepted). Even the samples don't show how to check the response for success.

When an error occurs, an appropriately typed exception should be thrown, so that users don't incorrectly assume the request was successful. This would also make it easier to act on the error based on the type of the exception. Currently the user has to know the details of the API to interpret the response. `SendGridClient` should abstract those details, otherwise it's just a thin wrapper around `HttpClient`. And doesn't really bring any value.

### LINKED ISSUE #796: Error Exception Handling
#### Issue Summary

To improve the developer experience, we would like to pass back a [summary of the error received](https://sendgrid.com/docs/API_Reference/Web_API_v3/Mail/errors.html) and a link to the appropriate place [in the documentation](https://sendgrid.com/docs/API_Reference/Web_API_v3/Mail/errors.html) to help with troubleshooting. 

#### Acceptance Criteria
 

* When an error is thrown due to a failed request, a summary of that error along with a [link to the appropriate place in the documentation](https://sendgrid.com/docs/API_Reference/Web_API_v3/Mail/errors.html) should be sent back as a JSON object.
* Tests that demonstrate successful functionality

### PR DESCRIPTION
**Pull request by childish-sambino**
Replaces #801 after addressing merge conflicts
Fixes #796 
Fixes #775

### CODE REVIEW

**Comment by thinkingserious** [comment]
I believe this one also resolves #774.

**Comment by thinkingserious** [comment]
I think this also solves #854.

File: src/SendGrid/Helpers/Errors/Model/RequestErrorException.cs [hunk 1]
```diff
@@ -0,0 +1,37 @@
+﻿using System;
+
+namespace SendGrid.Helpers.Errors.Model
+{
+    /// <summary>
+    /// Represents errors with status code 4xx. Being x any number
+    /// </summary>
+    public class RequestErrorException : Exception
```
Comment (eshanholtz) [inline src/SendGrid/Helpers/Errors/Model/RequestErrorException.cs:8]: Not sure I understand why we have custom exceptions for 403, 404, 405, 413, etc. as well as a 4xx catch-all exception. Is this just for more granularity on commonly thrown 4xx exceptions?

File: src/SendGrid/Helpers/Errors/Model/SendGridInternalException.cs [hunk 2]
```diff
@@ -0,0 +1,37 @@
+﻿using System;
+
+namespace SendGrid.Helpers.Errors.Model
+{
+    /// <summary>
+    /// Represents errors with status code 5xx. Being x any number
```
Comment (eshanholtz) [inline src/SendGrid/Helpers/Errors/Model/SendGridInternalException.cs:6]: Same question as above, w.r.t 5xx exceptions

File: src/SendGrid/Helpers/Errors/Model/RequestErrorException.cs (hunk 1 above)
Comment (childish-sambino) [reply src/SendGrid/Helpers/Errors/Model/RequestErrorException.cs:8]: Yes, that's correct.

I'm making a change now so the more specific 4xx and 5xx exceptions inherit from these bases.

**Pull request merged**